### PR TITLE
Recycle port HLD

### PR DESCRIPTION
**Pull request by ysmanman**

### CODE REVIEW

**Review by eswaranb — COMMENTED** [review]
1. Support host if for recycle port?
2. Update show command to support the recycle port.
3. Inband port can be configured to be Recirc0

File: doc/voq/recycle_port.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+Two recycle ports, Recirc0 and Recirc1, are configured in the example port_config.ini below. The port role is Rec, which means Recirc0 and Recirc1 will recycle packet to be routed.
+
+```
+#name         lanes                     alias        index  role       speed
+Ethernet0     48,49,50,51,52,53,54,55   Ethernet1/1  1      Ext        400000
+Ethernet8     56,57,58,59,60,61,62,63   Ethernet2/1  2      Ext        400000
+Ethernet16    64,65,66,67,68,69,70,71   Ethernet3/1  3      Ext        400000
+Ethernet24    72,73,74,75,76,77,78,79   Ethernet4/1  4      Ext        400000
+Recirc0       221                       Recirc0/0    5      Rec        400000
```
Comment (eswaranb) [inline doc/voq/recycle_port.md:75]: Inband0       221                      Inband0    5      Rec        400000
Recirc0       222                      Recirc0    6      Rec        400000

File: doc/voq/recycle_port.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+Recirc1       222                       Recirc0/1    6      Rec        400000
```
Comment (ysmanman) [inline doc/voq/recycle_port.md:76]: Add a new role Inb for inband.

File: doc/voq/recycle_port.md (hunk 1 above)
Comment (ysmanman) [inline doc/voq/recycle_port.md:75]: Rename Recirc0 -> Ethernet-Recirc0.

**Comment by anshuv-mfst** [comment]
@rlhui , @arlakshm - Could you please review the PR, thanks.  
@skeesara-nokia - could you please review the PR, thanks.

**Review by rlhui — COMMENTED** [review]
How to create a recycle port with SAI?

**Comment by ysmanman** [comment]
> How to create a recycle port with SAI?

This should be similar to front panel ports. Recycle ports are defined in port_config.ini, and are populated in APP_DB port table. portsorch retrieves ports list, including recycle ports, from SAI in initialization, and initialize recycle ports just like other front panel ports in doTask.

**Comment by rlhui** [comment]
> > How to create a recycle port with SAI?
> 
> This should be similar to front panel ports. Recycle ports are defined in port_config.ini, and are populated in APP_DB port table. portsorch retrieves ports list, including recycle ports, from SAI in initialization, and initialize recycle ports just like other front panel ports in doTask.

how to convey to ASIC SAI implementation that this is not an actual front-panel port, but to add 1 or 2 recycle ports?

**Comment by ysmanman** [comment]
> > > How to create a recycle port with SAI?
> > 
> > 
> > This should be similar to front panel ports. Recycle ports are defined in port_config.ini, and are populated in APP_DB port table. portsorch retrieves ports list, including recycle ports, from SAI in initialization, and initialize recycle ports just like other front panel ports in doTask.
> 
> how to convey to ASIC SAI implementation that this is not an actual front-panel port, but to add 1 or 2 recycle ports?

If a SAI implementation needs to differentiate recycle port from recycle ports, it can encode a sub port type in the pid of recycle port. But this is transparent to orchagent since all orchagent needs is just a pid for recycle port.

**Comment by abdosi** [comment]
Some Question on QoS w.r.t Recycle Port:-

What about the Queue of recycle port ? Do we will have Queue stats for recycle port ?  What will be behaviour of PFC and recycle port ? Can this port get back pressure for PFC queue 3 and 4 ?

**Review by abdosi — CHANGES_REQUESTED** [review]
Some questions on Qos.

**Comment by skeesara-nokia** [comment]
> 
> 
> Some Question on QoS w.r.t Recycle Port:-
> 
> What about the Queue of recycle port ? Do we will have Queue stats for recycle port ? What will be behaviour of PFC and recycle port ? Can this port get back pressure for PFC queue 3 and 4 ?



> 
> 
> Some Question on QoS w.r.t Recycle Port:-
> 
> What about the Queue of recycle port ? Do we will have Queue stats for recycle port ? What will be behaviour of PFC and recycle port ? Can this port get back pressure for PFC queue 3 and 4 ?

Recycle port is expected to have queue statistics like any other port.
The back pressure part may be implementation dependent. Expect more clarity on that once the Recycle port proposal is reviewed in the SAI forum.

**Comment by skeesara-nokia** [comment]
> 
> 
> Some questions on Qos.

QOS for the recycle port in general works similar any other port. It will support queues, mappings, remarking etc.

File: doc/voq/recycle_port.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+Two recycle ports, Ethernet-Rec0 and Ethernet-Rec1, are configured in the example port_config.ini below. Ethernet-Rec0 is used by Everflow, which has port role Rec. Ethernet-Rec1 is used as an inband port and thus its port role is set to Inb. The recycle port's lanes, used to discover the corresponding SAI ports, must be provided in port_config.ini as well.
+
+```
+#name               lanes                     alias        index  role       speed
+Ethernet0           48,49,50,51,52,53,54,55   Ethernet1/1  1      Ext        400000
+Ethernet8           56,57,58,59,60,61,62,63   Ethernet2/1  2      Ext        400000
+Ethernet16          64,65,66,67,68,69,70,71   Ethernet3/1  3      Ext        400000
+Ethernet24          72,73,74,75,76,77,78,79   Ethernet4/1  4      Ext        400000
+Ethernet-Rec0       221                       Recirc0/0    5      Rec        400000
+Ethernet-Rec1       222                       Recirc0/1    6      Inb        400000
```
[the remark this answers is not in the document]
Comment (ysmanman) [reply doc/voq/recycle_port.md]: Sounds good. Fixed.

**Review by skeesara-nokia — APPROVED** [review]

**Comment by rlhui** [comment]
> _No description provided._



> > Some Question on QoS w.r.t Recycle Port:-
> > What about the Queue of recycle port ? Do we will have Queue stats for recycle port ? What will be behaviour of PFC and recycle port ? Can this port get back pressure for PFC queue 3 and 4 ?
> 
> > Some Question on QoS w.r.t Recycle Port:-
> > What about the Queue of recycle port ? Do we will have Queue stats for recycle port ? What will be behaviour of PFC and recycle port ? Can this port get back pressure for PFC queue 3 and 4 ?
> 
> Recycle port is expected to have queue statistics like any other port.
> The back pressure part may be implementation dependent. Expect more clarity on that once the Recycle port proposal is reviewed in the SAI forum.

I would like to request that we add to the doc the usage of the SAI recycle port API - the current raised idea is to add a subport type for the recycle port.  How will this be used by the sonic applications/users?

**Comment by ysmanman** [comment]
> > _No description provided._
> 
> > > Some Question on QoS w.r.t Recycle Port:-
> > > What about the Queue of recycle port ? Do we will have Queue stats for recycle port ? What will be behaviour of PFC and recycle port ? Can this port get back pressure for PFC queue 3 and 4 ?
> > 
> > 
> > > Some Question on QoS w.r.t Recycle Port:-
> > > What about the Queue of recycle port ? Do we will have Queue stats for recycle port ? What will be behaviour of PFC and recycle port ? Can this port get back pressure for PFC queue 3 and 4 ?
> > 
> > 
> > Recycle port is expected to have queue statistics like any other port.
> > The back pressure part may be implementation dependent. Expect more clarity on that once the Recycle port proposal is reviewed in the SAI forum.
> 
> I would like to request that we add to the doc the usage of the SAI recycle port API - the current raised idea is to add a subport type for the recycle port. How will this be used by the sonic applications/users?

The HLD proposes to use recycle port in a way that’s independent of the SAI recycele port API. In this proposal, SONiC application/users shouldn’t really care about if a subport type is defined for recycle port or not. What we proposed in the HLD is to use port role to identify recycle ports in SONiC and its applications. Detailedly, we define recycle ports by setting the specific port roles along with their lanes in port_config.ini. With lanes and roles from port_config.ini, portsorch should be able to initialize recycle ports and also uncover their pid returned by SAI. SONiC applications that need recycle ports, like Everflow, should simply request the pid of recycle port from portsorch. It’s not clear to me why we have to define a subport type for recycle port. Did I get your comment correct?

**Comment by ysmanman** [comment]
> > > > > How to create a recycle port with SAI?
> > > > 
> > > > 
> > > > This should be similar to front panel ports. Recycle ports are defined in port_config.ini, and are populated in APP_DB port table. portsorch retrieves ports list, including recycle ports, from SAI in initialization, and initialize recycle ports just like other front panel ports in doTask.
> > > 
> > > 
> > > how to convey to ASIC SAI implementation that this is not an actual front-panel port, but to add 1 or 2 recycle ports?
> > 
> > 
> > If a SAI implementation needs to differentiate recycle port from recycle ports, it can encode a sub port type in the pid of recycle port. But this is transparent to orchagent since all orchagent needs is just a pid for recycle port.
> 
> > > > > How to create a recycle port with SAI?
> > > > 
> > > > 
> > > > This should be similar to front panel ports. Recycle ports are defined in port_config.ini, and are populated in APP_DB port table. portsorch retrieves ports list, including recycle ports, from SAI in initialization, and initialize recycle ports just like other front panel ports in doTask.
> > > 
> > > 
> > > how to convey to ASIC SAI implementation that this is not an actual front-panel port, but to add 1 or 2 recycle ports?
> > 
> > 
> > If a SAI implementation needs to differentiate recycle port from recycle ports, it can encode a sub port type in the pid of recycle port. But this is transparent to orchagent since all orchagent needs is just a pid for recycle port.
> 
> Would we please add to the doc on this. Thanks.

As I replied to another comment,  this seems not relevant to the proposal in the HLD.

**Comment by ysmanman** [comment]
> Some questions on Qos.

This comment had been addressed already. Copy the reply again to update review status:

QOS for the recycle port in general works similar any other port. It will support queues, mappings, remarking etc.

**Review by abdosi — APPROVED** [review]

**Pull request merged**